File: tools/census.py
"""The corpus census, and the gate on it (RFD 2236, step 2 precondition).

Reads every stage directory given, counts rows per family, template and frame, the
block kinds each rank1 program uses, the program length histogram, the tokens per
row in the text form under Gemma 4's tokenizer, and the two holdout axes: a family
or a block kind named as held out must appear in no training row. A leak is a
refusal; `--self-test` plants one and must be refused.

    python tools/census.py work/stage work/stage_react work/stage_harness --holdout-blocks TOF,RS,GE,MOD
"""
from __future__ import annotations

import argparse
import collections
import json
import re
import statistics
import sys
from pathlib import Path

import pyarrow.parquet as pq

BLOCK_RE = re.compile(r"^\w+ = ([A-Z_]+)(?:\[|\()", re.M)


def blocks_of(text: str) -> list[str]:
    return BLOCK_RE.findall(text)


def rows_of(stage: Path, split_dir: Path) -> list[dict]:
    fam = json.loads((stage / "manifest.json").read_text(encoding="utf-8")).get("family", "fbd")
    root = pq.read_table(split_dir / f"{fam}_root").to_pylist()
    cands = pq.read_table(split_dir / f"{fam}_candidates").to_pylist()
    ref = {c["row_key"]: c for c in cands if c["candidate"] == "rank1"}
    out = []
    for r in root:
        c = ref[r["key"]]
        path = stage / (c.get("fbd_text") or c.get("fbd_xml"))
        text = path.read_text(encoding="utf-8")
        if path.suffix == ".xml":
            text = re.sub(r'typeName="([A-Z_]+)"', lambda m: f"\n_ = {m.group(1)}(", text)
        out.append({"family": fam, "template": r["template_id"], "frame": r.get("frame_id", 0),
                    "blocks": blocks_of(text), "text": text})
    return out


def tokens(texts: list[str]) -> list[int] | None:
    try:
        from huggingface_hub import hf_hub_download
        from tokenizers import Tokenizer
        tok = Tokenizer.from_file(hf_hub_download("google/gemma-4-E2B-it-qat-q4_0-unquantized", "tokenizer.json"))
    except Exception as e:  # noqa: BLE001
        print(f"tokens: not counted ({e.__class__.__name__}); pass --no-tokens to silence", file=sys.stderr)
        return None
    return [len(tok.encode(t).ids) for t in texts]


def census(stages: list[Path], holdout_families: set[str], holdout_blocks: set[str], count_tokens: bool) -> tuple[dict, list[str]]:
    train, holdout = [], []
    for s in stages:
        train += rows_of(s, s / "data")
        holdout += rows_of(s, s / "holdout" / "data")
    problems = []
    fam_counts = collections.Counter(r["family"] for r in train)
    tpl_counts = collections.Counter(r["template"] for r in train)
    block_counts = collections.Counter(b for r in train for b in set(r["blocks"]))
    lengths = collections.Counter(len(r["blocks"]) for r in train)
    for r in train:
        if r["template"] in holdout_families:
            problems.append(f"held-out family {r['template']} leaked into training")
            break
    leaked = {b for r in train for b in r["blocks"] if b in holdout_blocks}
    for b in sorted(leaked):
        problems.append(f"held-out block kind {b} leaked into training")
    for b in sorted(holdout_blocks):
        if not any(b in r["blocks"] for r in holdout):
            problems.append(f"held-out block kind {b} appears in no holdout row either; the axis is empty")
    tok = tokens([r["text"] for r in train]) if count_tokens else None
    report = {
        "train_rows": len(train), "holdout_rows": len(holdout),
        "families": dict(fam_counts), "templates": dict(sorted(tpl_counts.items())),
        "frames": dict(collections.Counter(f"{r['template']}#{r['frame']}" for r in train)),
        "block_kinds": dict(sorted(block_counts.items())),
        "block_kinds_present": len(block_counts),
        "length_histogram": dict(sorted(lengths.items())),
        "holdout_families": sorted(holdout_families), "holdout_blocks": sorted(holdout_blocks),
        "tokens": None if tok is None else {"mean": round(statistics.mean(tok), 1), "p50": statistics.median(tok),
                                             "max": max(tok), "total": sum(tok)},
    }
    return report, problems


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("stages", nargs="+", type=Path)
    ap.add_argument("--holdout-families", default="")
    ap.add_argument("--holdout-blocks", default="")
    ap.add_argument("--no-tokens", action="store_true")
    ap.add_argument("--min-rows", type=int, default=0)
    ap.add_argument("--self-test", action="store_true")
    ap.add_argument("--out", type=Path, default=None)
    args = ap.parse_args()
    hf = {f for f in args.holdout_families.split(",") if f}
    hb = {b for b in args.holdout_blocks.split(",") if b}
    report, problems = census(args.stages, hf, hb, not args.no_tokens)
    if args.self_test:
        # plant a leak: the most common block kind declared held out
        planted = next(iter(sorted(report["block_kinds"], key=report["block_kinds"].get, reverse=True)))
        _, planted_problems = census(args.stages, hf, hb | {planted}, False)
        if not any("leaked" in p for p in planted_problems):
            sys.exit(f"self-test: planted held-out block {planted} was not refused; the census is decoration")
        print(f"self-test: planted {planted} refused")
    if args.min_rows and report["train_rows"] < args.min_rows:
        problems.append(f"{report['train_rows']} train rows, below the floor of {args.min_rows}")
    print(json.dumps(report, indent=2))
    if args.out:
        args.out.write_text(json.dumps(report, indent=2), encoding="utf-8")
    for p in problems:
        print("FAIL", p)
    if problems:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: lib/taskweft_fbd_teacher/census.ex
defmodule TaskweftFbdTeacher.Census do
  @moduledoc """
  The corpus census and the gate on it. Counts rows per family, template and frame, the
  block kinds each rank1 program uses, the program length histogram, and the two holdout
  axes: a family or a block kind named as held out must appear in no training row.

  Block kinds are read by splitting rather than by pattern, because the operator's ruling
  is to parse where parsing is possible and because `fbd.no_regex` is a gate this module
  has to pass.
  """

  alias Explorer.DataFrame, as: DF
  alias Explorer.Series

  @upper ~c"ABCDEFGHIJKLMNOPQRSTUVWXYZ_"
  @sig ~c"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_."

  @doc "Block kinds in a text-form program, a CALL counted by its signature."
  def block_kinds(text) do
    Enum.reject(assignments(text), &(&1 == "CALL")) ++ call_signatures(text)
  end

  defp assignments(text) do
    text
    |> String.split("\n")
    |> Enum.flat_map(fn line ->
      case String.split(line, " = ", parts: 2) do
        [lhs, rhs] -> if word?(lhs), do: opening_kind(rhs), else: []
        _ -> []
      end
    end)
  end

  # The kind is an upper-case run immediately followed by its argument bracket, which is
  # what separates `b3 = WRITE_FILE(...)` from an assignment of a plain name.
  defp opening_kind(rhs) do
    {kind, rest} = Enum.split_while(String.to_charlist(rhs), &(&1 in @upper))

    if kind != [] and match?([c | _] when c in [?[, ?(], rest),
      do: [List.to_string(kind)],
      else: []
    end

  defp word?(s), do: s != "" and String.to_charlist(s) |> Enum.all?(&(&1 in ~c"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_"))

  defp call_signatures(text), do: call_signatures(text, [])

  defp call_signatures(text, acc) do
    case String.split(text, "CALL[", parts: 2) do
      [_] ->
        Enum.reverse(acc)

      [_, rest] ->
        {sig, tail} = Enum.split_while(String.to_charlist(rest), &(&1 in @sig))
        sig = List.to_string(sig)

        acc =
          if match?([?] | _], tail) and String.contains?(sig, "."), do: [sig | acc], else: acc

        call_signatures(List.to_string(tail), acc)
    end
  end

  @doc "rank1 rows of one split, with the program text each names."
  def rows_of(stage, split_dir) do
    family = family_of(stage)
    root = read_table(Path.join(split_dir, "#{family}_root"))
    cands = read_table(Path.join(split_dir, "#{family}_candidates"))

    if root == nil or cands == nil do
      []
    else
      rank1 =
        cands
        |> Enum.filter(&(&1["candidate"] == "rank1"))
        |> Map.new(&{&1["row_key"], &1})

      for r <- root, c = rank1[r["key"]], c != nil do
        text = c["fbd_text"] || c["fbd_xml"] || ""

        %{
          family: family,
          template: r["template_id"],
          frame: r["frame_id"] || 0,
          blocks: block_kinds(text)
        }
      end
    end
  end

  defp family_of(stage) do
    Path.join(stage, "manifest.json")
    |> File.read!()
    |> :json.decode()
    |> Map.get("family", "fbd")
  end

  defp read_table(dir) do
    file =
      cond do
        File.dir?(dir) -> dir |> Path.join("*.parquet") |> Path.wildcard() |> List.first()
        File.regular?(dir) -> dir
        true -> nil
      end

    with path when is_binary(path) <- file, {:ok, df} <- DF.from_parquet(path) do
      names = DF.names(df)
      cols = Map.new(names, &{&1, df |> DF.pull(&1) |> Series.to_list()})
      for i <- 0..(DF.n_rows(df) - 1)//1, do: Map.new(names, &{&1, Enum.at(cols[&1], i)})
    else
      _ -> nil
    end
  end

  @doc """
  The report and the problems. A held-out family or block kind appearing in a training
  row is a problem, and so is a holdout axis that no held-out row exercises, because an
  empty axis reads as a pass.
  """
  def run(stages, holdout_families \\ MapSet.new(), holdout_blocks \\ MapSet.new()) do
    train = Enum.flat_map(stages, &rows_of(&1, Path.join(&1, "data")))
    test = Enum.flat_map(stages, &rows_of(&1, Path.join([&1, "test", "data"])))

    held =
      Enum.flat_map(stages, fn s ->
        Enum.flat_map(["holdout", "evaluation"], &rows_of(s, Path.join([s, &1, "data"])))
      end)

    leaked_family = Enum.any?(train, &MapSet.member?(holdout_families, &1.template))

    leaked_blocks =
      train
      |> Enum.flat_map(& &1.blocks)
      |> Enum.filter(&MapSet.member?(holdout_blocks, &1))
      |> Enum.uniq()
      |> Enum.sort()

    empty_axes =
      holdout_blocks
      |> Enum.sort()
      |> Enum.reject(fn b -> Enum.any?(held, &(b in &1.blocks)) end)

    problems =
      (if leaked_family, do: ["a held-out family leaked into training"], else: []) ++
        Enum.map(leaked_blocks, &"held-out block kind #{&1} leaked into training") ++
        Enum.map(empty_axes, &"held-out block kind #{&1} appears in no holdout row either; the axis is empty")

    report = %{
      "train_rows" => length(train),
      "test_rows" => length(test),
      "evaluation_rows" => length(held),
      "families" => tally(train, & &1.family),
      "templates" => tally(train, & &1.template),
      "frames" => tally(train, &"#{&1.template}##{&1.frame}"),
      "block_kinds" => tally(train, & &1.blocks, :flat),
      "length_histogram" => tally(train, &to_string(length(&1.blocks))),
      "holdout_families" => Enum.sort(holdout_families),
      "holdout_blocks" => Enum.sort(holdout_blocks)
    }

    {Map.put(report, "block_kinds_present", map_size(report["block_kinds"])), problems}
  end

  # A block kind is counted once per row, matching the set the Python census takes.
  defp tally(rows, fun, :flat),
    do: rows |> Enum.flat_map(&(&1 |> fun.() |> Enum.uniq())) |> Enum.frequencies()

  defp tally(rows, fun), do: rows |> Enum.map(fun) |> Enum.frequencies()

  @doc "Plants the most common block kind as held out; the census must refuse it."
  def self_test(stages) do
    {report, _} = run(stages)

    case Enum.max_by(report["block_kinds"], &elem(&1, 1), fn -> nil end) do
      nil ->
        {:error, "no block kinds counted; there is nothing for the control to plant"}

      {kind, _} ->
        {_, problems} = run(stages, MapSet.new(), MapSet.new([kind]))

        if Enum.any?(problems, &String.contains?(&1, "leaked")),
          do: {:ok, kind},
          else: {:error, "planted #{kind} was not refused; the census is decoration"}
    end
  end
end
